src/Spanner/Session/CacheSessionPool.php

```diff
@@ -0,0 +1,794 @@
+            $data['inUse'][$session->name()]['lastActive'] = $this->time();
+
+            $this->cacheItemPool->save($item->set($data));
+        });
+    }
+
+    /**
+     * Downsizes the queue of available sessions by the given percentage. This is
+     * relative to the minimum sessions value. For example: Assuming a full
+     * queue, with maximum sessions of 10 and a minimum of 2, downsizing by 50%
+     * would leave 6 sessions in the queue. The count of items to be deleted will
+     * be rounded up in the case of a fraction.
+     *
+     * A session may be removed from the cache, but still tracked as active by
+     * the Spanner backend if a delete operation failed. To ensure you do not
+     * exceed the maximum number of sessions available per node, please be sure
+     * to check the return value of this method to be certain all sessions have
+     * been deleted.
+     *
+     * Please note this method will attempt to synchronously delete sessions and
+     * will block until complete.
+     *
+     * @param int $percent The percentage to downsize the pool by. Must be
+     *        between 1 and 100.
+     * @return array An associative array containing a key `deleted` which holds
+     *         an integer value representing the number of queued sessions
+     *         deleted on the backend and a key `failed` which holds a list of
+     *         queued {@see Google\Cloud\Spanner\Session\Session} objects which
+     *         failed to delete.
+     * @throws \InvaldArgumentException
+     */
+    public function downsize($percent)
+    {
+        if ($percent < 1 || 100 < $percent) {
+            throw new \InvalidArgumentException('The provided percent must be between 1 and 100.');
+        }
+
+        $failed = [];
+        $toDelete = $this->config['lock']->synchronize(function () use ($percent) {
+            $item = $this->cacheItemPool->getItem($this->cacheKey);
+            $data = (array) $item->get() ?: $this->initialize();
+            $toDelete = [];
+            $queueCount = count($data['queue']);
+            $availableCount = max($queueCount - $this->config['minSessions'], 0);
+            $countToDelete = ceil($availableCount * ($percent * 0.01));
+
+            if ($countToDelete) {
+                $toDelete = array_splice($data['queue'], (int) -$countToDelete);
+            }
+
+            $this->cacheItemPool->save($item->set($data));
+            return $toDelete;
+        });
+
+        foreach ($toDelete as $sessionData) {
+            $session = $this->database->session($sessionData['name']);
+
+            try {
+                $session->delete();
+            } catch (\Exception $ex) {
+                if ($ex instanceof NotFoundException) {
+                    continue;
+                }
+
+                $failed[] = $session;
+            }
+        }
+
+        return [
+            'deleted' => count($toDelete) - count($failed),
+            'failed' => $failed
```

Comment:
The thought here was in the unlikely case a delete operation fails, a user should have the ability to make sure these sessions are removed manually at their own discretion. The cache itself is completely cleared, so if these references to the sessions aren't retained there will be no way to delete the sessions until they expire.